src/NUnitFramework/framework/Internal/TestParameters.cs

```diff
@@ -116,30 +116,42 @@ private void InitializeArguments(object[] args)
         /// </summary>
         public string TestName { get; set; }
 
+        /// <summary>
+        /// If not <see langword="null"/>, overrides the list of argument display names to be used when generating a test name.
+        /// </summary>
+        // TODO: use private protected as soon as the codebase supports C# 7.2
+        internal string[] ArgDisplayNames { private get; set; }
+
         /// <summary>
         /// Gets the property dictionary for this test
         /// </summary>
         public IPropertyBag Properties { get; private set; }
 
         #endregion
 
-        #region IApplyToTest Members
-
         /// <summary>
-        /// Applies ParameterSet values to the test itself.
+        /// Applies the encapsulated parameters to the test.
         /// </summary>
-        /// <param name="test">A test.</param>
-        public void ApplyToTest(Test test)
+        public void ApplyToTest(Test test, TestNameGenerator defaultTestNameGenerator)
         {
-            if (this.RunState != RunState.Runnable)
-                test.RunState = this.RunState;
+            if (RunState != RunState.Runnable)
+                test.RunState = RunState;
 
             foreach (string key in Properties.Keys)
                 foreach (object value in Properties[key])
                     test.Properties.Add(key, value);
-        }
 
-        #endregion
+            if (TestName == null)
+            {
+                test.Name = defaultTestNameGenerator.GetDisplayName(test, ArgDisplayNames, ArgDisplayNames != null ? null : OriginalArguments);
+            }
+            else
+            {
+                test.Name = TestName.Contains("{")
+                    ? new TestNameGenerator(TestName).GetDisplayName(test, ArgDisplayNames, ArgDisplayNames != null ? null : OriginalArguments)
+                    : TestName;
+            }
+        }
```

Comment:
I think this is the key bit I find unexpected. Why do TestParameters have responsibility for setting a test name? It feels more natural to me that the Case/FixtureBuilders calll TestNameGenerator, and pass in TestParams.

Reply:
Recap from https://github.com/nunit/nunit/pull/2538#discussion_r156253485: it allows TestParameters to keep its private state private and decide how to apply it. Do we really want to make both `SetArgDisplayNames` and `ArgDisplayNames` public? If another name-generation-state field is added to TestParameters, do we really want to add them in a pair? ... Maybe. It wasn't my first choice. I like the ApplyTo pattern.